sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders/ExpressionEncoder.scala

```diff
@@ -162,6 +163,18 @@ object ExpressionEncoder {
       e4: ExpressionEncoder[T4],
       e5: ExpressionEncoder[T5]): ExpressionEncoder[(T1, T2, T3, T4, T5)] =
     tuple(Seq(e1, e2, e3, e4, e5)).asInstanceOf[ExpressionEncoder[(T1, T2, T3, T4, T5)]]
+
+  /**
+   * Function that deserializes an [[InternalRow]] into an object of type `T`. Instances of this
+   * class are not meant to be thread-safe.
+   */
+  abstract class Deserializer[T] extends (InternalRow => T)
```

Comment:
Implementation note. I opted to go with abstract classes so we can get monomorphic call sites in many cases.